linkerd/proxy/api-resolve/src/resolve.rs

```diff
@@ -88,30 +85,35 @@ where
     fn call(&mut self, target: T) -> Self::Future {
         let path = target.to_string();
         debug!(dst = %path, context = %self.context_token, "Resolving");
-        let mut svc = self.service.clone();
         let req = api::GetDestination {
             path,
             scheme: self.scheme.clone(),
             context_token: self.context_token.clone(),
         };
-        Box::pin(async move {
-            let rsp = svc.get(grpc::Request::new(req)).await?;
-            trace!(metadata = ?rsp.metadata());
-            Ok(Resolution {
-                inner: rsp.into_inner(),
-            })
-        })
+
+        future::ok(Box::pin(resolution(self.service.clone(), req)))
     }
 }
 
-impl Stream for Resolution {
-    type Item = Result<resolve::Update<Metadata>, grpc::Status>;
-
-    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
-        let mut this = self.project();
-        loop {
-            match ready!(this.inner.as_mut().poll_next(cx)) {
-                Some(update) => match update?.update {
+fn resolution<S>(
+    mut client: DestinationClient<S>,
+    req: api::GetDestination,
+) -> impl Stream<Item = Result<resolve::Update<Metadata>, grpc::Status>>
+where
+    S: GrpcService<BoxBody> + Clone + Send + 'static,
+    S::Error: Into<Box<dyn Error + Send + Sync>> + Send,
+    S::ResponseBody: Send,
+    <S::ResponseBody as Body>::Data: Send,
+    <S::ResponseBody as HttpBody>::Error: Into<Box<dyn Error + Send + Sync + 'static>> + Send,
+    S::Future: Send,
+{
+    try_stream! {
+            let rsp = client.get(grpc::Request::new(req)).await?;
+            trace!(metadata = ?rsp.metadata());
+            let mut stream = rsp.into_inner();
+    loop {
+            while let Some(update) = stream.next().await {
```

Comment:
We definitely don't want to continue looping after the stream returns `None`... so this needs to turn into `loop { match ...`

Reply:
But `loop / match` hits the same error... so Something is not properly handling the end-of stream...